File: packages/i18nify-js/src/modules/paymentTranslations/__tests__/getPaymentTranslationsList.test.ts
import getPaymentTranslationsList from '../getPaymentTranslationsList';
import { getPaymentTranslationsInfo } from '../utils';
import { PAYMENT_TRANSLATIONS_CODE_LIST } from '../constants';
import type { PaymentTranslationsStrings } from '../types';

const EXPECTED_KEYS: (keyof PaymentTranslationsStrings)[] = [
  'language_name',
  'script',
  'pay_now',
  'amount',
  'cancel',
  'confirm',
  'payment_successful',
  'payment_failed',
  'processing',
  'total',
  'retry',
  'enter_amount',
  'card_number',
  'card_expiry',
  'cvv',
  'name_on_card',
  'upi_id',
  'net_banking',
  'wallet',
  'transaction_id',
  'back',
  'error_try_again',
  'order_id',
  'receipt',
];

describe('getPaymentTranslationsList', () => {
  describe('return shape', () => {
    it('returns an object', () => {
      const result = getPaymentTranslationsList();
      expect(typeof result).toBe('object');
      expect(result).not.toBeNull();
    });

    it('contains exactly the 7 Indian language codes', () => {
      const result = getPaymentTranslationsList();
      const keys = Object.keys(result).sort();
      expect(keys).toEqual([...PAYMENT_TRANSLATIONS_CODE_LIST].sort());
    });

    it('every entry has all 24 required string fields', () => {
      const result = getPaymentTranslationsList();
      for (const code of PAYMENT_TRANSLATIONS_CODE_LIST) {
        const entry = result[code];
        for (const key of EXPECTED_KEYS) {
          expect(entry).toHaveProperty(key);
          expect(typeof entry[key]).toBe('string');
        }
      }
    });

    it('no entry has empty strings for core fields', () => {
      const result = getPaymentTranslationsList();
      const coreFields: (keyof PaymentTranslationsStrings)[] = [
        'language_name',
        'script',
        'pay_now',
        'amount',
        'cancel',
        'confirm',
      ];
      for (const code of PAYMENT_TRANSLATIONS_CODE_LIST) {
        for (const field of coreFields) {
          expect(result[code][field]).not.toBe('');
        }
      }
    });
  });

  describe('Hindi (hi)', () => {
    it('has correct language_name and script', () => {
      const result = getPaymentTranslationsList();
      expect(result['hi'].language_name).toBe('Hindi');
      expect(result['hi'].script).toBe('Devanagari');
    });

    it('pay_now is non-empty', () => {
      expect(getPaymentTranslationsList()['hi'].pay_now).toBeTruthy();
    });
  });

  describe('Bengali (bn)', () => {
    it('has correct language_name and script', () => {
      const result = getPaymentTranslationsList();
      expect(result['bn'].language_name).toBe('Bengali');
      expect(result['bn'].script).toBe('Bengali');
    });
  });

  describe('Tamil (ta)', () => {
    it('has correct language_name and script', () => {
      const result = getPaymentTranslationsList();
      expect(result['ta'].language_name).toBe('Tamil');
      expect(result['ta'].script).toBe('Tamil');
    });
  });

  describe('Telugu (te)', () => {
    it('has correct language_name and script', () => {
      const result = getPaymentTranslationsList();
      expect(result['te'].language_name).toBe('Telugu');
      expect(result['te'].script).toBe('Telugu');
    });
  });

  describe('Kannada (kn)', () => {
    it('has correct language_name and script', () => {
      const result = getPaymentTranslationsList();
      expect(result['kn'].language_name).toBe('Kannada');
      expect(result['kn'].script).toBe('Kannada');
    });
  });

  describe('Gujarati (gu)', () => {
    it('has correct language_name and script', () => {
      const result = getPaymentTranslationsList();
      expect(result['gu'].language_name).toBe('Gujarati');
      expect(result['gu'].script).toBe('Gujarati');
    });
  });

  describe('Marathi (mr)', () => {
    it('has correct language_name and script', () => {
      const result = getPaymentTranslationsList();
      expect(result['mr'].language_name).toBe('Marathi');
      expect(result['mr'].script).toBe('Devanagari');
    });
  });

  describe('all languages produce distinct pay_now strings', () => {
    it('each language has a unique pay_now translation', () => {
      const result = getPaymentTranslationsList();
      const payNowValues = PAYMENT_TRANSLATIONS_CODE_LIST.map(
        (code) => result[code].pay_now,
      );
      const unique = new Set(payNowValues);
      expect(unique.size).toBe(PAYMENT_TRANSLATIONS_CODE_LIST.length);
    });
  });
});

describe('getPaymentTranslationsInfo (utils)', () => {
  it('returns strings for a valid code', () => {
    const result = getPaymentTranslationsInfo('hi');
    expect(result).not.toBeNull();
    expect(result?.language_name).toBe('Hindi');
  });

  it('returns null for an unknown code', () => {
    const result = getPaymentTranslationsInfo('xx' as never);
    expect(result).toBeNull();
  });

  it('returns all expected keys for Tamil', () => {
    const result = getPaymentTranslationsInfo('ta');
    expect(result).not.toBeNull();
    for (const key of EXPECTED_KEYS) {
      expect(result).toHaveProperty(key);
    }
  });
});

File: packages/i18nify-go/modules/payment_translations/get_payment_translations_list_test.go
package payment_translations

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

var expectedLanguageCodes = []string{"hi", "bn", "mr", "gu", "ta", "te", "kn"}

func TestGetPaymentTranslationsList_NoError(t *testing.T) {
	result, err := GetPaymentTranslationsList()
	require.NoError(t, err)
	require.NotNil(t, result)
}

func TestGetPaymentTranslationsList_AllSevenLanguages(t *testing.T) {
	result, err := GetPaymentTranslationsList()
	require.NoError(t, err)

	for _, code := range expectedLanguageCodes {
		_, ok := result[code]
		assert.True(t, ok, "expected language code %q to be present", code)
	}
	assert.Len(t, result, len(expectedLanguageCodes))
}

func TestGetPaymentTranslationsList_LanguageMetadata(t *testing.T) {
	tests := []struct {
		code         string
		languageName string
		script       string
	}{
		{"hi", "Hindi", "Devanagari"},
		{"bn", "Bengali", "Bengali"},
		{"mr", "Marathi", "Devanagari"},
		{"gu", "Gujarati", "Gujarati"},
		{"ta", "Tamil", "Tamil"},
		{"te", "Telugu", "Telugu"},
		{"kn", "Kannada", "Kannada"},
	}

	result, err := GetPaymentTranslationsList()
	require.NoError(t, err)

	for _, tt := range tests {
		t.Run(tt.code, func(t *testing.T) {
			entry, ok := result[tt.code]
			require.True(t, ok, "code %q not found", tt.code)
			assert.Equal(t, tt.languageName, entry.LanguageName)
			assert.Equal(t, tt.script, entry.Script)
		})
	}
}

func TestGetPaymentTranslationsList_AllFieldsNonEmpty(t *testing.T) {
	result, err := GetPaymentTranslationsList()
	require.NoError(t, err)

	for _, code := range expectedLanguageCodes {
		entry := result[code]
		t.Run(code, func(t *testing.T) {
			assert.NotEmpty(t, entry.LanguageName, "language_name empty for %s", code)
			assert.NotEmpty(t, entry.Script, "script empty for %s", code)
			assert.NotEmpty(t, entry.PayNow, "pay_now empty for %s", code)
			assert.NotEmpty(t, entry.Amount, "amount empty for %s", code)
			assert.NotEmpty(t, entry.Cancel, "cancel empty for %s", code)
			assert.NotEmpty(t, entry.Confirm, "confirm empty for %s", code)
			assert.NotEmpty(t, entry.PaymentSuccessful, "payment_successful empty for %s", code)
			assert.NotEmpty(t, entry.PaymentFailed, "payment_failed empty for %s", code)
			assert.NotEmpty(t, entry.Processing, "processing empty for %s", code)
			assert.NotEmpty(t, entry.Total, "total empty for %s", code)
			assert.NotEmpty(t, entry.Retry, "retry empty for %s", code)
			assert.NotEmpty(t, entry.EnterAmount, "enter_amount empty for %s", code)
			assert.NotEmpty(t, entry.CardNumber, "card_number empty for %s", code)
			assert.NotEmpty(t, entry.CardExpiry, "card_expiry empty for %s", code)
			assert.NotEmpty(t, entry.CVV, "cvv empty for %s", code)
			assert.NotEmpty(t, entry.NameOnCard, "name_on_card empty for %s", code)
			assert.NotEmpty(t, entry.UPIID, "upi_id empty for %s", code)
			assert.NotEmpty(t, entry.NetBanking, "net_banking empty for %s", code)
			assert.NotEmpty(t, entry.Wallet, "wallet empty for %s", code)
			assert.NotEmpty(t, entry.TransactionID, "transaction_id empty for %s", code)
			assert.NotEmpty(t, entry.Back, "back empty for %s", code)
			assert.NotEmpty(t, entry.ErrorTryAgain, "error_try_again empty for %s", code)
			assert.NotEmpty(t, entry.OrderID, "order_id empty for %s", code)
			assert.NotEmpty(t, entry.Receipt, "receipt empty for %s", code)
		})
	}
}

func TestGetPaymentTranslationsList_UniquePayNowStrings(t *testing.T) {
	result, err := GetPaymentTranslationsList()
	require.NoError(t, err)

	seen := make(map[string]string)
	for _, code := range expectedLanguageCodes {
		payNow := result[code].PayNow
		if prev, exists := seen[payNow]; exists {
			t.Errorf("duplicate pay_now value %q for codes %s and %s", payNow, prev, code)
		}
		seen[payNow] = code
	}
}

func TestGetPaymentTranslationsList_Idempotent(t *testing.T) {
	r1, err1 := GetPaymentTranslationsList()
	r2, err2 := GetPaymentTranslationsList()
	require.NoError(t, err1)
	require.NoError(t, err2)
	assert.Equal(t, r1, r2)
}
